File: benchmarks/expr_eval.js
const { performance } = require('perf_hooks');

function num(n) { return { kind: 0, n: n }; }
function add(a, b) { return { kind: 1, a: a, b: b }; }
function mul(a, b) { return { kind: 2, a: a, b: b }; }

function evaluate(e) {
    if (e.kind === 0) return e.n;
    if (e.kind === 1) return evaluate(e.a) + evaluate(e.b);
    return evaluate(e.a) * evaluate(e.b);
}

function build(depth) {
    if (depth <= 0) return num(1);
    return add(mul(build(depth - 1), num(2)), build(depth - 1));
}

const t0 = performance.now();
let total = 0;
for (let i = 0; i < 10; i++) {
    total += evaluate(build(16));
}
console.log(total);
console.log(`bench_ns=${Math.round((performance.now() - t0) * 1e6)}`);

File: benchmarks/expr_eval.rs
use std::time::Instant;

enum Expr {
    Num(i32),
    Add(Box<Expr>, Box<Expr>),
    Mul(Box<Expr>, Box<Expr>),
}

fn num(n: i32) -> Box<Expr> { Box::new(Expr::Num(n)) }
fn add(a: Box<Expr>, b: Box<Expr>) -> Box<Expr> { Box::new(Expr::Add(a, b)) }
fn mul(a: Box<Expr>, b: Box<Expr>) -> Box<Expr> { Box::new(Expr::Mul(a, b)) }

fn eval(e: &Expr) -> i32 {
    match e {
        Expr::Num(n) => *n,
        Expr::Add(a, b) => eval(a) + eval(b),
        Expr::Mul(a, b) => eval(a) * eval(b),
    }
}

fn build(depth: i32) -> Box<Expr> {
    if depth <= 0 {
        return num(1);
    }
    add(mul(build(depth - 1), num(2)), build(depth - 1))
}

fn main() {
    let t0 = Instant::now();
    let mut total = 0;
    for _ in 0..10 {
        total += eval(&build(16));
    }
    println!("{}", total);
    println!("bench_ns={}", t0.elapsed().as_nanos());
}
